lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,534 @@
+        # with the column name we will abort the update request.
+        non_existent = list(set(post_data.keys()) - set(User.__table__.columns.keys()))
+        non_updatable = list(
+            set(post_data.keys()).intersection(set(User.get_protected()))
+        )
+        if non_updatable:
+            for field in non_updatable:
+                if getattr(user, field) != post_data[f"{field}"]:
+                    self.logger.warning(
+                        "Either provided values to update the user does not adhere to the user model "
+                        "datatype or user attempted to update the protected properties."
+                    )
+                    abort(400, message="Invalid data in update request payload")
+        if non_existent:
+            self.logger.warning(
+                "User trying to update fields that are not present in the user database. Fields: {}",
+                non_existent,
+            )
+            abort(400, message="Invalid fields in update request payload")
+        try:
+            user.update(**post_data)
+        except Exception:
+            self.logger.exception("Exception occurred during updating user object")
+            abort(500, message="INTERNAL ERROR")
+
+        return_data = user.get_json()
+        response_object = {
+            "status": "success",
+            "data": return_data,
+        }
+        return make_response(jsonify(response_object), 200)
+
+    @Auth.token_auth.login_required(f=Auth().verify_auth())
+    def delete(self, username):
+        """
+        Delete request for deleting a user from database.
+        This requires a Pbench auth token in the header field
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+            Authorization:   Bearer Pbench_auth_token (user received upon login)
+
+        :return: JSON Payload
+            response_object = {
+                    "status": "success"/"fail",
+                    "status_code": 200/status_code_int,
+                    "message": "Successfully deleted."/"failure message",
+                }
+        """
+        try:
+            user, verified = self.auth.verify_user(username)
+        except Exception:
+            self.logger.exception("Exception occurred during the getUser {}", username)
+            abort(500, message="INTERNAL ERROR")
+
+        # TODO: Check if the user has the right privileges
+        if not verified and not user.is_admin():
+            self.logger.warning(
+                "Username retrieved from the auth token {} is different from the username provided",
+                self.auth.auth.current_user().id,
+            )
+            abort(403, message="Authentication token does not belong to the user")
+
+        try:
+            if not User.delete(username):
+                self.logger.warning("Admin attempted to delete admin user")
```

Comment:
Apart from the mechanism/policy question, I don't understand why l.519 means what the message says it means: how can we conclude that that is the case? I'm probably missing something obvious...

Reply:
Fixed, @ndokos you are right it was a bit ambiguous, I have updated it

Reply:
@npalaska, the new code is much better -- thank you.

@ndokos, the old code was assuming that, if the attempt to delete failed, it was because the target was the admin user, because the delete mechanism was also enforcing the policy.  The new code, which places the policy enforcement in the caller, is much clearer.